lib/json/string_parser.nit

```diff
@@ -233,30 +233,98 @@ class JSONStringParser
 
 	# Parses and returns a Nit string from a JSON String
 	fun parse_json_string: Jsonable do
+		var src = src
 		var ln = src.length
 		var p = pos
 		p += 1
 		if p > ln then return make_parse_error("Malformed JSON String")
+		var ret: nullable Buffer = null
 		var c = src[p]
-		var st = p
+		var chunk_st = p
 		while c != '"' do
-			if c == '\\' then
-				if p + 1 >= ln then return make_parse_error("Malformed Escape sequence in JSON string")
+			if c != '\\' then
 				p += 1
+				if p >= ln then return make_parse_error("Malformed JSON string")
 				c = src[p]
-				if c == 'u' then
+				continue
+			end
+			if ret == null then ret = new Buffer
+			ret.append_substring(src, chunk_st, p - chunk_st)
+			p += 1
+			if p >= ln then return make_parse_error("Malformed Escape sequence in JSON string")
+			c = src[p]
+			if c == 'r' then
+				ret.add '\r'
+				p += 1
+			else if c == 'n' then
+				ret.add '\n'
+				p += 1
+			else if c == 't' then
+				ret.add '\t'
+				p += 1
+			else if c == 'u' then
+				var cp = 0
+				p += 1
+				for i in [0 .. 4[ do
+					cp <<= 4
+					if p >= ln then make_parse_error("Malformed \uXXXX Escape sequence in JSON string")
+					c = src[p]
+					if c >= '0' and c <= '9' then
+						cp += c.code_point - '0'.code_point
+					else if c >= 'a' and c <= 'f' then
+						cp += c.code_point - 'a'.code_point + 10
+					else if c >= 'A' and c <= 'F' then
+						cp += c.code_point - 'A'.code_point + 10
+					else
+						make_parse_error("Malformed \uXXXX Escape sequence in JSON string")
+					end
 					p += 1
-					if p + 3 >= ln then return make_parse_error("Bad Unicode escape sequence in string")
-					for i in [0 .. 4[ do if not src[p + i].is_hexdigit then return make_parse_error("Bad Unicode escape sequence in string")
-					p += 3
 				end
+				c = cp.code_point
+				if c.is_hi_surrogate then
+					if p >= ln then make_parse_error("Malformed \uXXXX Escape sequence in JSON string")
+					c = src[p]
+					if c != '\\' then make_parse_error("Malformed \uXXXX Escape sequence in JSON string")
+					p += 1
+					c = src[p]
+					if c != 'u' then make_parse_error("Malformed \uXXXX Escape sequence in JSON string")
```

Comment:
Yup, that's the kind of thing to do if you want to handle `\uXXXX` escape sequences since these are UTF-16, hence there's a possibility they are surrogate pairs.
If the first \uXXXX read was a surrogate pair, then the next char must be a low-surrogate pair, which must be escaped since it is an illegal UTF-8 character.